### ISSUE TITLE
Add support for PodDisruptionBudgets

### ISSUE THREAD

**Issue by JorTurFer** [open-telemetry/opentelemetry-operator#2136]
Although `sidecar` and `daemonset` don't allow pdb, `deployment` and `statefulset` modes are workload types where pdbs make sense totally, as they are centralized workloads. 
Currently, we are using the collector helm chart because it supports pdbs but we'd like to use the operator for using `Target Allocator`, but the pdbs are a requirement for us.

I think that adding support to pdbs is a good addition to the operator and I'm willing to contribute with the feature if you also agree with having it

**Comment by pavolloffay** [comment]
thanks for the request @JorTurFer 

Do you want to use `PodDisruptionBudgets` for target allocator as well?

**Comment by JorTurFer** [comment]
I haven't tested properly the target allocator yet, but if it's something that requires HA, yes. I mean, I know that we can use affinities to prevent some scenarios, but we have to ensure that we have at least X instances of the collector working. That's why we needed the pdbs

**Comment by JorTurFer** [comment]
Do you see the request useful? I'm willing to implement (but obviously, I prefer to ask before rather than drafting a PR that never is merged xD)

**Comment by pavolloffay** [comment]
I am not well familiar with the concept perhaps @jaronoff97 ?

I found this 

> From version 1.15 PDBs support custom controllers where the [scale subresource](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#scale-subresource) is enabled.

https://kubernetes.io/docs/tasks/run-application/configure-pdb/#identify-an-application-to-protect

otelcol CRD implements the scale subresource

**Comment by JorTurFer** [comment]
No no, I want to deploy a pdb for the generated deployment / statefulset. I mean, with this manifest:
```yaml
apiVersion: opentelemetry.io/v1alpha1
kind: OpenTelemetryCollector
metadata:
  name: sidecar-for-my-app
spec:
  mode: deployment
```
The operator generates a deployment for the collector. I want to generate also a PDB for that generated deployment, not for the CRD itself

**Comment by jaronoff97** [comment]
Hey @JorTurFer thanks for getting this going here :) Is there a reason we would do this in the operator and not just let users apply this themselves (that's what i do currently)? What pavol mentions is accurate in that if we were to do this in the operator we would want to do it the same way as how the hpa created by the operator works i.e. by targeting the scale subresource.

**Comment by jaronoff97** [comment]
yeah makes sense! I think as long as we let it work with the collector in the same way as the HPA that would be great. I definitely would accept it!

**Comment by JorTurFer** [comment]
Nice!
Let me take a look and give a try. You can assign the issue to me if you want :)

**Comment by jaronoff97** [comment]
sounds great, thanks for taking this on :bow:

**Comment by JorTurFer** [comment]
I think that I have already something to review :)

**Comment by cmergenthaler** [comment]
Hey @JorTurFer, thanks for adding the feature! Is there a reason why the PDB hasn't been added for `Target Allocator` as well?

**Comment by JorTurFer** [comment]
As I understood the TargetAllocator isn't critical and doesn't require HA (but maybe I'm wrong and we have to add it there too)
I mean, it does not process real time traffic but configure the collector, so 5-10 seconds down are not critical

**Comment by cmergenthaler** [comment]
Thanks for the explanation. From what I've understood, it is recommended to run at least 2 replicas of TA, so that it is also HA, see https://github.com/open-telemetry/opentelemetry-operator/issues/2159#issuecomment-1739785276
So I think having pdb in place would be valid as well

**Comment by jaronoff97** [comment]
@cmergenthaler i think that's a valid concern – would you mind opening up another issue? I think there's reason here to do some more work to pipe these things better for user concerns. I'll add a discussion topic for our SIG meeting this week.

**Comment by cmergenthaler** [comment]
@jaronoff97 done: https://github.com/open-telemetry/opentelemetry-operator/issues/2261

**Comment by JorTurFer** [comment]
Oh, I misunderstood the TA I guess. 
Let's continue the discussion on the issue :)
